lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -5663,92 +5474,515 @@ class Differentiation : public SILModuleTransform {
 };
 } // end anonymous namespace
 
+SILFunction *
+ADContext::getOrCreateLinearMapIndexSubsetThunk(
+    SILFunction *parentThunk, CanSILFunctionType linearMapType,
+    CanSILFunctionType targetType, AutoDiffAssociatedFunctionKind kind,
+    SILAutoDiffIndices desiredIndices, SILAutoDiffIndices actualIndices) {
+  auto &astCtx = getASTContext();
+
+  SubstitutionMap interfaceSubs = parentThunk->getForwardingSubstitutionMap();
+  GenericEnvironment *genericEnv = parentThunk->getGenericEnvironment();
+  auto thunkType = buildThunkType(
+      parentThunk, linearMapType, targetType, genericEnv, interfaceSubs,
+      /*withoutActuallyEscaping*/ true,
+      DifferentiationThunkKind::Reabstraction);
+
+  // TODO: Use more principled mangling.
+  std::string thunkName;
+  switch (kind) {
+    case AutoDiffAssociatedFunctionKind::JVP:
+      thunkName = "differential";
+      break;
+    case AutoDiffAssociatedFunctionKind::VJP:
+      thunkName = "pullback";
+  }
+  Mangle::ASTMangler mangler;
+  auto fromInterfaceType =
+      linearMapType->mapTypeOutOfContext()->getCanonicalType();
+  auto toInterfaceType = targetType->mapTypeOutOfContext()->getCanonicalType();
+  CanType dynamicSelfType;
+  thunkName = "AD__" + mangler.mangleReabstractionThunkHelper(
+      thunkType, fromInterfaceType, toInterfaceType, dynamicSelfType,
+      module.getSwiftModule()) + "_" + desiredIndices.mangle() + "_" +
+      thunkName;
+  thunkName += "_thunk";
+
+  auto loc = parentThunk->getLocation();
+  SILOptFunctionBuilder fb(getTransform());
+  auto *thunk = fb.getOrCreateSharedFunction(
+      loc, thunkName, thunkType, IsBare, IsTransparent, IsSerialized,
+      ProfileCounter(), IsThunk, IsNotDynamic);
+
+  if (!thunk->empty())
+    return thunk;
+
+  thunk->setGenericEnvironment(genericEnv);
+  thunk->setOwnershipEliminated();
+  auto *entry = thunk->createBasicBlock();
+  SILBuilder builder(entry);
+  createEntryArguments(thunk);
+
+  // Get arguments.
+  SmallVector<SILValue, 4> arguments;
+  SmallVector<AllocStackInst *, 4> localAllocations;
+
+  // Build a `.zero` argument for the given `Differentiable`-conforming type.
+  auto buildZeroArgument = [&](SILType zeroSILType) {
```

Comment:
This logic should be unified with existing logic used for `emitZeroIndirect` instead of being duplicated.

Reply:
Done in https://github.com/apple/swift/pull/24845/commits/646096e004e5525b1baefd17701bd4f07a55ad40.